### ISSUE TITLE
Receiving "Must contain an `adapter` key referencing the adapter to use." errors during Heroku deployment

### ISSUE THREAD

**Issue by dniccum** [balderdashy/sails#3499]
I am in the process of deploying a basic SailsJS application to Heroku. I have done this a few times before without any hiccups; which leads me to believe this might be a documentation or deprecation issue.

When I deploy the application, this is the error that I am getting over and over:

```
Must contain an `adapter` key referencing the adapter to use.
In model (class), invalid connection :: mongoProd
```

I am currently using the following versions:

```
sails: 0.11.4
sails-mongo: 0.11.6
```

Now, I can confirm that have the `sails-mongo` adapter installed (duh). As far as my production.js is concerned, this is what I have:

```
environment: 'production',
  hookTimeout: 30000,

  // blueprint settings
  blueprints: {
    rest: false,
    shortcuts: false
  },

  session: {
    adapter: 'redis',
    host: //,
    port: 9592,
    user: //,
    pass: //,
  },
  models: {
    connection: 'mongoProd',
    migrate: 'alter'
  },
  port: process.env.PORT || 5000,

  log: {
    level: "silent"
  }
```

And finally, this is what my connections.js looks like:

```
mongoProd: {
    adapter: 'sails-mongo',
    url: process.env.MONGOLAB_URI
  },
```

I feel like Sails is not finding any of my environment configurations.

**Comment by japel** [comment]
are you able to lift locally? if so, can you lift in console mode and do a console.log(sails.config.connections); and post the output?
also console.log(sails.config.connections); and post the output?

**Comment by mikermcneil** [comment]
@japel great questions
@dniccum you might already know this, but please be very, very careful using `migrate: alter` in production!  Part 2 of [this](https://courses.platzi.com/courses/develop-apps-sails-js/) covers doing a deployment to Heroku and might be helpful here.

**Comment by dniccum** [comment]
@japel I can lift in my local environment using `sails lift`, however when I use `sails lift --prod`, this is the error that I get: 

```
error: Error: Failed to connect to MongoDB.  Are you sure your configured Mongo instance is running?
Error details:
{ [MongoError: connect ECONNREFUSED 127.0.0.1:27017]
  name: 'MongoError',
  message: 'connect ECONNREFUSED 127.0.0.1:27017' }
    at _createError (/Users/dougniccum/Sites/quiq-math-backend/node_modules/sails-mongo/lib/adapter.js:103:23)
    at /Users/dougniccum/Sites/quiq-math-backend/node_modules/sails-mongo/lib/adapter.js:106:13
    at /Users/dougniccum/Sites/quiq-math-backend/node_modules/sails-mongo/lib/connection.js:25:20
    at /Users/dougniccum/Sites/quiq-math-backend/node_modules/sails-mongo/node_modules/mongodb/lib/mongo_client.js:276:20
    at /Users/dougniccum/Sites/quiq-math-backend/node_modules/sails-mongo/node_modules/mongodb/lib/db.js:224:14
    at null.<anonymous> (/Users/dougniccum/Sites/quiq-math-backend/node_modules/sails-mongo/node_modules/mongodb/lib/server.js:234:9)
    at g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at emit (events.js:172:7)
    at null.<anonymous> (/Users/dougniccum/Sites/quiq-math-backend/node_modules/sails-mongo/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:269:68)
    at g (events.js:260:16)
    at emitTwo (events.js:87:13)
    at emit (events.js:172:7)
    at null.<anonymous> (/Users/dougniccum/Sites/quiq-math-backend/node_modules/sails-mongo/node_modules/mongodb/node_modules/mongodb-core/lib/connection/pool.js:81:12)
    at g (events.js:260:16)
    at emitTwo (events.js:87:13)
```

I am assuming this is normal since I am not running MongoDB locally. I am using my localDiskDB.db.

@mikermcneil noted!

**Comment by japel** [comment]
@dniccum your local mongodb is running on the specified host and port? (127.0.0.1:27017)

**Comment by japel** [comment]
@dniccum yeah, thats normal... try lifting locally in console mode with localDiskDB and do those console.log() I wrote, just to check if those are defined

**Comment by dniccum** [comment]
@japel This is what is returned when I run the `console.log(sails.config.connections)` command:

```
{ localDiskDb: { adapter: 'sails-disk' },
  someMysqlServer:
   { adapter: 'sails-mysql',
     host: 'YOUR_MYSQL_SERVER_HOSTNAME_OR_IP_ADDRESS',
     user: 'YOUR_MYSQL_USER',
     password: 'YOUR_MYSQL_PASSWORD',
     database: 'YOUR_MYSQL_DB' },
  mongoProd: { adapter: 'sails-mongo', url: undefined },
  somePostgresqlServer:
   { adapter: 'sails-postgresql',
     host: 'YOUR_POSTGRES_SERVER_HOSTNAME_OR_IP_ADDRESS',
     user: 'YOUR_POSTGRES_USER',
     password: 'YOUR_POSTGRES_PASSWORD',
     database: 'YOUR_POSTGRES_DB' } }
```

Once again, I'm assuming that the reason why the url attribute is undefined is because the environment variable has not been defined. As for the `console.log(sails.config.environment)` command, this is what I get:

```
development
```

**Comment by mikermcneil** [comment]
@dniccum how do things look if you start up your Sails app using:

```bash
NODE_ENV=production node app
```

**Comment by dniccum** [comment]
@mikermcneil Nothing happens.

**Comment by japel** [comment]
@dniccum which operating system are you using?

**Comment by dniccum** [comment]
OSX El Capitan.

**Comment by japel** [comment]
@dniccum and what do you mean with nothing happens?
what is the output of your terminal when doing
NODE_ENV=production node app.js
?

**Comment by dniccum** [comment]
Yes when I enter this into my terminal, I receive no response or additional
messages.

On Mon, Jan 18, 2016 at 11:37 AM Jan Viktor Apel <notifications@github.com>
wrote:

> @dniccum <https://github.com/dniccum> and what do you mean with nothing
> happens?
> what is the output of your terminal when doing
> NODE_ENV=production node app.js
> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balderdashy/sails/issues/3499#issuecomment-172601366>.
>
-- 
Doug Niccum

Sent from my iPhone. Sorry if I have fat fingers.

**Comment by mikermcneil** [comment]
@dniccum Would you mind setting up a repo that reproduces this in a new sails app real quick?  I think that'd be the easiest way to see what's going on.

e.g.:
```bash
sails new sails-issue-3499
cd sails-issue-3499
npm install sails-mongo@0.11.6 --save
# Now make whichever are the simplest/quickest changes which reproduce the issue,
# and when you for sure have the issue isolated, run:
git init
hub create  # or use the GitHub UI to create the repo and `git remote add` it manually
git add .
git commit -am 'Reproduces https://github.com/balderdashy/sails/issues/3499'
git push -u origin master
hub browse # or open the repo in your browser manually
# Then copy the URL of the new repo and paste it as a comment on this issue.
# We'll check it out, and since this way we'll be able to experience the problem
# ourselves, we should have no problem getting to the bottom of it.
```

Thanks man!

**Comment by dniccum** [comment]
After some toying around, I have determined that the problem had to do with the version of redis-connect that I was using. Once I reduced it to lock in version 1.4.5, the application works just fine. Sorry for wasting your time.

**Comment by mikermcneil** [comment]
@dniccum Not at all - sorry for the confusion with versions.  I tried my best w/ the error messages, but it's sort of an uphill battle with the peer dependencies.  Luckily, I think we've got session store dependency versioning [more or less figured out](https://github.com/balderdashy/sails/issues/3486#issuecomment-171111185) for the upcoming v0.12 release, so getting the right version should get easier soon.
